### PR TITLE
Namespace compatibility changes

### PR DESCRIPTION
**Pull request by rochacbruno**
Trying to rework #1437 

https://issues.redhat.com/browse/AAH-2293

### CODE REVIEW

File: pulp_ansible/app/custom_fields.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+from pulpcore.plugin.util import (
+    assign_role,
+    remove_role,
+    get_groups_with_perms_attached_roles,
+    get_perms_for_model,
+)
+from pulpcore.plugin.models import Group
+
+
+class RelatedFieldsBaseSerializer(serializers.Serializer):
```
Comment (mdellweg) [inline pulp_ansible/app/custom_fields.py:24]: Is this a reinvention of drf_queryfields?

File: pulp_ansible/app/custom_fields.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            # When there is no request present, it usually means that the serializer is
+            # being inspected by DRF spectacular to generate the open API spec. In that
+            # case, this should act like a normal serializer.
+            return super().to_representation(instance)
+
+        return result
+
+
+class MyPermissionsField(serializers.Serializer):
+    def to_representation(self, original_obj):
```
Comment (mdellweg) [inline pulp_ansible/app/custom_fields.py:78]: This has tremendous resemblance with the `my_permissions` action in our `pulpcore.app.viewsets.base.RolesMixin`.

```python
    @extend_schema(
        summary="List user permissions",
        description="List permissions available to the current user on this object.",
        responses={
            200: inline_serializer(
                name="MyPermissionsSerializer", fields={"permissions": ListField(child=CharField())}
            )
        },
    )
    @action(detail=True, methods=["get"])
    def my_permissions(self, request, pk=None):
        obj = self.get_object()
        app_label = obj._meta.app_label
        permissions = [
            ".".join((app_label, codename)) for codename in request.user.get_all_permissions(obj)
        ]
        return Response({"permissions": permissions})
```

File: pulp_ansible/app/custom_fields.py (hunk 1 above)
Comment (rochacbruno) [reply pulp_ansible/app/custom_fields.py:24]: I am not aware about drf_queryfields, but this is the pattern we use in other places in galaxy API

**Comment by rochacbruno** [comment]
Hi @mdellweg do you think any changes are required on this PR?

File: pulp_ansible/app/viewsets.py [hunk 3]
```diff
@@ -376,6 +391,16 @@ class Meta:
             "metadata_sha256": ["exact", "in"],
         }
 
+    def no_op(self, queryset, name, value):
+        return queryset
+
+    def has_permissions(self, queryset, name, value):
+        perms = self.request.query_params.getlist(name)
```
Comment (mdellweg) [inline pulp_ansible/app/viewsets.py:398]: How is that different from `value`?

File: pulp_ansible/app/viewsets.py [hunk 4]
```diff
@@ -368,6 +377,12 @@ class AnsibleNamespaceFilter(ContentFilter):
     A filter for namespaces.
     """
 
+    # this is a hack to keep pulp from throwing a 400 when an unknwon
+    # query param is included, since filtering is handled by the related
+    # field serializer
+    include_related = filters.CharFilter(method="no_op")
+    my_permissions = filters.CharFilter(method="has_permissions")
+
```
Comment (mdellweg) [inline pulp_ansible/app/viewsets.py:385]: This part I don't get. These are pulp api views.
Also we don't break down permissions to the content. I believe we should have the `RolesMixin` in the `AnsibleGlobalNamespaceViewSet` instead.

File: pulp_ansible/app/galaxy/v3/views.py [hunk 5]
```diff
@@ -799,50 +798,34 @@ def _create(self, request, data, context=None):
...
-                    artifact = Artifact.objects.get(sha256=artifact.sha256)
-                    artifact.touch()
-                except (Artifact.DoesNotExist, DatabaseError):
-                    # the artifact has since been removed from when we first attempted to save it
-                    artifact.save()
-            context["artifact"] = artifact.pk
-
-        # Dispatch general_create task
-        app_label = AnsibleNamespaceMetadata._meta.app_label
+        c = serializer.save()
```
Comment (mdellweg) [inline pulp_ansible/app/galaxy/v3/views.py:801]: `c` is not such a good variable name.

File: pulp_ansible/app/custom_fields.py (hunk 2 above)
Comment (mdellweg) [reply pulp_ansible/app/custom_fields.py:78]: Can you at least consider using

```python
        permissions = [
            ".".join((app_label, codename)) for codename in request.user.get_all_permissions(obj)
        ]
```

File: pulp_ansible/app/custom_fields.py (hunk 1 above)
Comment (mdellweg) [reply pulp_ansible/app/custom_fields.py:24]: So this is a bit of code copied from galaxy_ng?
Let me rephrase: Why is it called `include_related`? All I see is, it can select fields on the serializer to be included or not. And that is exactly what drf_queryfields (a pulp dependency for about forever) does.

File: pulp_ansible/app/serializers.py [hunk 6]
```diff
@@ -808,11 +818,38 @@ def to_representation(self, value):
...
+
+
+class AnsibleNamespaceMetadataSerializer(NoArtifactContentSerializer, NamespaceValidationMixin):
     """
     A serializer for Namespaces.
     """
 
+    # Task on the serializer will always return none. It's just here to prevent the bindings
+    # from breaking.
+    task = serializers.SerializerMethodField()
```
Comment (mdellweg) [inline pulp_ansible/app/serializers.py:850]: Why would the bindings break?
We generate them for each version individually anyway.
I'd rather not do such things.

**Review by mdellweg — COMMENTED** [review]
I left some suggestions/questions.
Still not quite confident to understand every aspect of this large PR.

**Comment by rochacbruno** [comment]
@newswangerd can you help me answer @mdellweg questions as you are the original author of this code?

File: pulp_ansible/app/custom_fields.py (hunk 1 above)
Comment (rochacbruno) [reply pulp_ansible/app/custom_fields.py:24]: @mdellweg I talked to David and here is the explanation:

include_related controls what is included in the specific field that is based on RelatedFieldsSerializer, this is not to control the regular fields from the serializer, but to control which fields are included nested in a specialized field, the name is `include_related` because it is always data coming from related models.

So instead of always fetching the related models we do that only when is explicitly specified in the query param.

File: pulp_ansible/app/custom_fields.py (hunk 2 above)
Comment (rochacbruno) [reply pulp_ansible/app/custom_fields.py:78]: It is implemented this way to be included in the ListView, while the one from RolesMixin is for DEtailView only

File: pulp_ansible/app/viewsets.py (hunk 3 above)
Comment (rochacbruno) [reply pulp_ansible/app/viewsets.py:398]: Looks like it needs to be taken using getlist() so it returns a Python list type.

File: pulp_ansible/app/viewsets.py (hunk 4 above)
Comment (rochacbruno) [reply pulp_ansible/app/viewsets.py:385]: Those fields must be declared, otherwise we are not able to pass those fields as query params.

File: pulp_ansible/tests/functional/api/collection/v3/test_collection_version_search.py [hunk 7]
```diff
@@ -1458,8 +1458,7 @@ def test_cross_repo_search_index_with_updated_namespace_metadata(
     task = galaxy_v3_plugin_namespaces_api_client.create(
         name=namespace_name, description="hello", company="Testing Co.", **distro_kwargs
     )
-    result = monitor_task(task.task)
-    namespace_href = [x for x in result.created_resources if "namespaces" in x][0]
+    namespace_href = task.pulp_href
```
Comment (mdellweg) [inline pulp_ansible/tests/functional/api/collection/v3/test_collection_version_search.py:1461]: Is this api no longer using tasks? Then we should rename `task` to `namespace`.

File: pulp_ansible/tests/functional/api/collection/v3/test_collection_version_search.py [hunk 8]
```diff
@@ -1489,8 +1488,7 @@ def test_cross_repo_search_index_with_updated_namespace_metadata(
         avatar=avatar_path,
         **distro_kwargs,
     )
-    result2 = monitor_task(task2.task)
-    namespace2_href = [x for x in result2.created_resources if "namespaces" in x][0]
+    namespace2_href = task2.pulp_href
```
Comment (mdellweg) [inline pulp_ansible/tests/functional/api/collection/v3/test_collection_version_search.py:1491]: Here too.

File: pulp_ansible/tests/functional/api/collection/v3/test_namespace.py [hunk 9]
```diff
@@ -41,6 +44,336 @@ def _random_image():
...
+        "2cool4school",
+        "x",
+        "xy",
+        "my-namespace-name",
+        "my.namespace.name",
+        "hello!",
+    ]
+
+    for name in invalid_names:
+        try:
```
Comment (mdellweg) [inline pulp_ansible/tests/functional/api/collection/v3/test_namespace.py:69]: This should use `with pytest.raises`

File: pulp_ansible/tests/functional/api/collection/v3/test_namespace.py [hunk 10]
```diff
@@ -41,6 +44,336 @@ def _random_image():
...
+):
+    """
+    Test that the legacy groups field works.
+    """
+
+    group_1 = random_string()
+    group_2 = random_string()
+
+    groups_api_client.create({"name": group_1})
+    groups_api_client.create({"name": group_2})
```
Comment (mdellweg) [inline pulp_ansible/tests/functional/api/collection/v3/test_namespace.py:102]: can we use gen_object_with_cleanup?

File: pulp_ansible/tests/functional/api/collection/v3/test_namespace.py [hunk 11]
```diff
@@ -41,6 +44,336 @@ def _random_image():
...
+
+    repo, distro = ansible_repo_and_distro_factory()
+    kwargs = {"path": distro.base_path, "distro_base_path": distro.base_path}
+
+    # Create namespace with no groups
+    name = random_string()
+    result = galaxy_v3_plugin_namespaces_api_client.create(name=name, **kwargs)
+    monitor_task(result.task)
```
Comment (mdellweg) [inline pulp_ansible/tests/functional/api/collection/v3/test_namespace.py:110]: I'm confused now. Is this a created with a task, or no?

Comment (mdellweg) [reply pulp_ansible/tests/functional/api/collection/v3/test_namespace.py:110]: Also we probably want it cleaned up.

**Review by mdellweg — COMMENTED** [review]
Still not a full review... Sorry.

File: pulp_ansible/tests/functional/api/collection/v3/test_namespace.py (hunk 9 above)
Comment (rochacbruno) [reply pulp_ansible/tests/functional/api/collection/v3/test_namespace.py:69]: pytest.raises allows only a single exception and this test is checking for ApiException and also ApiValueError.

I am not sure why (I have not written the test) but looks like the API can raise one of those errors, I guess it will depend on the input.

I dont see any advantage of using pytst.raises in this case.

Comment (mdellweg) [reply pulp_ansible/tests/functional/api/collection/v3/test_namespace.py:69]: Hmm, I'll trust you on that.
But for the record, i believe, pytest.raises is more readable and it marks a test failed, not errored.
Do we really not know which one Exception we are to expect here?

File: pulp_ansible/app/custom_fields.py [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+        request = self.context.get("request", None)
+        if request is None:
+            return []
+        user = request.user
+
+        if original_obj._meta.proxy:
+            obj = original_obj._meta.concrete_model.objects.get(pk=original_obj.pk)
+        else:
+            obj = original_obj
```
Comment (mdellweg) [inline pulp_ansible/app/custom_fields.py:87]: I'm pretty sure, in Pulp we use permissions on the proxy model is that different in Galaxy?

(I think the Group is the only proxy model we use, and we needed that for additional permissions).

File: pulp_ansible/app/custom_fields.py [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+                            "Role {role} does not exist or does not "
+                            "have any permissions related to this object."
+                        ).format(role=role)
+                    }
+                )
+
+
+class GroupFieldType(serializers.Serializer):
+    """
+    Group open api type
```
Comment (mdellweg) [inline pulp_ansible/app/custom_fields.py:127]: This is more a nested serializer than an openapi type.

File: pulp_ansible/app/custom_fields.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+        data = dictionary.get(self.field_name, [])
+
+        # The open api client sends data as a form request rather than json
+        # because of the avatar URL. It converts a list like
+        # [{"foo": "bar"}, {"bar": "foo"}] into "{'foo': 'bar'}, {'bar': 'foo'}"
+        # This is a best effort attempt to capture that data and convert it into
+        # valid JSON and then transform it into a list that the API can understand
+        if isinstance(data, str):
+            try:
+                data = f"[{data}]".replace("'", '"')
```
Comment (mdellweg) [inline pulp_ansible/app/custom_fields.py:147]:
```suggestion
                data = f"[{data.replace("'", '"')}]"
```
